utilities/transactions/write_prepared_txn_db.cc

```diff
@@ -49,6 +50,29 @@ Status WritePreparedTxnDB::Initialize(
   return s;
 }
 
+Status WritePreparedTxnDB::Write(const WriteOptions& opts, WriteBatch* updates,
+                                 bool skip_cc) {
+  // TODO(myabandeh): optimize for skip_cc
```

Comment:
Sorry for the noob questions since I'm not familiar with RocksDB code. What needs to happen here? Typically, does WritePreparedTxnDB wrap around PessimisticTransactionDB?  Should this do something for skip_cc being true, and then then call down to db_impl_->Write(opts, updates, skip_cc) vs. reimplementing PessimisticTransactionDB::Write()?

Reply:
WritePreparedTxnDB inherits from PessimisticTransactionDB. The skip_cc option is a hint that enabled optimization in the underlying TransactionDB. PessimisticTransactionDB's default behavior to make use of this optimization is to directly call DB::Write on the underlying DB. This implementation might or might not be correct for the classes that inherit from PessimisticTransactionDB. In the case of WritePreparedTxnDB this is not a correct behavior and ::Write needs to be overridden. The implementation here for WritePreparedTxnDB::Write is correct but i) inefficient as it does not make use of skip_cc hint, ii) has code overlap with PessimisticTransactionDB::Write as you pointed out.
The plan for this patch the minimal API changes that enables MyRocks to switch from DB::Write to TransactionDB::Write with no compromise on performance for the active write policy of WriteCommitted. There will be a future patch to also upgrade WritePrepared to make use of skip_cc hint (which would be more complicated) and also refactor the code to avoid code duplication (which would not change the exposed API). That future patch needs to more some logics from WritePreparedTxn to WritePreparedTxnDB so that it could implement ::Write without paying the overhead of beginning a new transaction (BeginInternalTransaction below).

Reply:
Ah, okay, that makes sense then. The current patch looks fine then. If you want something shorter, you could also have WritePreparedTxnDB::Write() to call PessimisticTransactionDB::Write(opts, update, false) directly and you can fill in the contents of WritePreparedTxnDB() later, rather than cutting/pasting code.